datafusion/optimizer/src/unwrap_cast_in_comparison.rs

```diff
@@ -140,140 +139,132 @@ struct UnwrapCastExprRewriter {
 impl TreeNodeRewriter for UnwrapCastExprRewriter {
     type Node = Expr;
 
-    fn f_up(&mut self, expr: Expr) -> Result<Transformed<Expr>> {
-        match &expr {
+    fn f_up(&mut self, mut expr: Expr) -> Result<Transformed<Expr>> {
+        match &mut expr {
```

Comment:
I was trying a few different patterns to remove `Expr` clones in this rule. The `match &mut expr` way seems to be most straightforward as `expr` is still available after pattern matching so we can return it unchanged if needed (`return Ok(Transformed::no(expr));`) but we can also mutate some parts of the `expr` if needed.